{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read fetching, organization, and preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This analysis uses the original file names for the sequencing runs. e.g. P3_H1_S10_R1_001.fastq.gz etc etc.\n",
    "To reproduce this analysis, first download the sequencing runs from NCBI Sequence Read Archive, and then rename accordingly:\n",
    "\n",
    "### OTTR sequencing data\n",
    "| SRA experiment ID | Original file name | Condition|\n",
    "| ---|---|----|\n",
    "| SRX25615775 | P12_A5_S6_L001_R1_001.fastq.gz | OTTR sequencing of DRT2 RNPs: WT\n",
    "| SRX25615777 | P12_A6_S7_L001_R1_001.fastq.gz | OTTR sequencing of DRT2 RNPs: YCAA mutant\n",
    "| SRX25615776 | P12_A7_S8_L001_R1_001.fastq.gz | OTTR sequencing of DRT2 RNPs: WT, DBR1 treated\n",
    "| SRX25615780 | P4_H11_S2_R1_001.fastq.gz | OTTR sequencing of DRT2 RNP + 16 uM dNTP reaction: alkaline hydrolysis treated\n",
    "| SRX25615778 | P7_H11_S1_R2_001.fastq.gz | OTTR sequencing of DRT2 RNPs: PBS mutant library\n",
    "| SRX25615779 | P7_H12_S2_R2_001.fastq.gz | OTTR sequencing of DRT2 RNP + 16 uM dNTP reaction: untreated\n",
    "\n",
    "Place renamed reads in folder OTTR_reads/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read trimming\n",
    "\n",
    "# phenol/chloroform extracted DRT2 RNPs. P12_A5 = WT. P12_A6 = YCDD>YCAA mutant. P12_A7 = WT treated with DRB1.\n",
    "# Sequenced on a MiSeq, R1 = 300bp.\n",
    "! cd OTTR_reads/\n",
    "! mkdir Trimmed\n",
    "! for input in P12_A*.fastq.gz; do out=Trimmed/$(echo $input | sed 's/fastq\\.gz/fastq/'); cutadapt --trim-n -q 20 -m 20 -a ATCGGTTCGAGATGGCTCG -g ^NNNNNR -o $out $input; done\n",
    "\n",
    "\n",
    "\n",
    "# phenol/chloroform extracted DRT2 RNPs where the PBS is a library of variants. P7_H11.\n",
    "# Sequenced on a MiSeq, R1 = 10 bp (discarded), R2 = 290 bp.\n",
    "! cutadapt --trim-n -q 20 -m 20 -a GYNNNNNCTGTCTCTTATACACATCTGACGCTGCCGACGA -g ^CGAGCCATCTCGAACCGATT -o Trimmed/P7_H11_S1_R2_001.fastq P7_H11_S1_R2_001.fastq.gz\n",
    "! cd Trimmed/\n",
    "! cutadapt -g ^CGAGCCATCTCGAACCGATT P7_H11_S1_R2_001.fastq > P7_H11_S1_R2_001_retrim.fastq\n",
    "! cd ../\n",
    "\n",
    "# DRT2 RNP incubated with 16 uM dNTPs, then phenol/chloroform extracted. P7_H12.\n",
    "# Sequenced on a MiSeq, R1 = 10 bp (discarded), R2 = 290 bp.\n",
    "! cutadapt --trim-n -q 20 -m 20 -g CGAGCCATCTCGAACCGATT -o Trimmed/H12_trim1.fastq P7_H12_S2_R2_001.fastq.gz \n",
    "! cd Trimmed/\n",
    "! cutadapt -g CGAGCCATCTCGAACCGATT -o H12_trim2.fastq H12_trim1.fastq\n",
    "! cutadapt -a CGAGCCATCTCGAACCGATT -o H12_trim3.fastq H12_trim2.fastq\n",
    "! cutadapt -a GYNNNNNCTGTCTCTTATACACATCT -m 20 -o H12_trim4.fastq H12_trim3.fastq\n",
    "! cd ../\n",
    "\n",
    "\n",
    "# DRT2 RNP incubated with 16 uM dNTPs, then RNA removed by alkaline hydrolysis, before phenol/chloroform extraction. P4_H11.\n",
    "# Sequenced on a MiSeq, R1 = 300 bp.\n",
    "! cutadapt --trim-n -q 20 -m 20 -a AATCGGTTCGAGATGGCTCG -o Trimmed/P4_H11_S2_R1_001.fastq P4_H11_S2_R1_001.fastq.gz\n",
    "! cd Trimmed\n",
    "! cutadapt -a AATCGGTTCGAGATGGCTCG -o H11_trim2.fastq P4_H11_S2_R1_001.fastq\n",
    "! cutadapt -a AATCGGTTCGAGATGGCTCG -m 8 -o H11_trim3.fastq H11_trim2.fastq \n",
    "! cutadapt -g ^NNNNNRC  -o H11_trim4.fastq H11_trim3.fastq \n",
    "! cd ../\n",
    "\n",
    "\n",
    "# read mapping\n",
    "! cd sam/\n",
    "! bwa index pLG128.fasta\n",
    "! bwa mem pLG128.fasta ../Trimmed/P7_H11_S1_R2_001_retrim.fastq > H11.sam\n",
    "# n.b. pLG128 is the in-house name for the plasmid pLG010 available on Addgene\n",
    "# all other reads were mapped using the Geneious read mapper, but similar results can be obtained using bwa mem as above"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# read analysis (in python)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run in bioconda environment\n",
    "\n",
    "#Import helpful packages\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['font.family'] = 'sans-serif'\n",
    "matplotlib.rcParams['font.sans-serif'] = 'Helvetica'\n",
    "matplotlib.rcParams['font.size'] = 10\n",
    "matplotlib.rcParams['axes.linewidth'] = 0.5\n",
    "matplotlib.rcParams['xtick.major.width'] = 0.5\n",
    "matplotlib.rcParams['ytick.major.width'] = 0.5\n",
    "\n",
    "matplotlib.rcParams['patch.force_edgecolor'] = True\n",
    "matplotlib.rcParams['patch.linewidth'] = 0.5\n",
    "\n",
    "\n",
    "#Packages for sequence parsing\n",
    "from Bio.Seq import Seq\n",
    "import pysam\n",
    "import logomaker\n",
    "\n",
    "#Some useful colors for the sequence logos\n",
    "\n",
    "logo_colorscheme = {\n",
    "    'A': \"#C63F59\",\n",
    "    'C': \"#B0492E\",\n",
    "    'G': \"#F5D9AF\",\n",
    "    'T': \"#E686AC\"\n",
    "}\n",
    "\n",
    "hanpink = \"#DC3860\"\n",
    "hanpink_dark = \"#C63F59\"\n",
    "hanpink_light = \"#F1617A\"\n",
    "hanpink_lightest = \"#E686AC\"\n",
    "hanpinkpalette = [hanpink_dark, hanpink, hanpink_light, hanpink_lightest]\n",
    "\n",
    "darkgrey = \"#737880\"\n",
    "midgrey = \"#949598\"\n",
    "lightgrey = \"#C8CECE\"\n",
    "lightestgrey = \"#DBDAD5\"\n",
    "greypalette = [darkgrey, midgrey, lightgrey, lightestgrey]\n",
    "drtpalette = [\"#B0492E\", \"#EB6E4A\", \"#E39625\", \"#F5D9AF\"]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 4D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this cell creates a frequency matrix for plotting for the experiment comparing WT DRT RNPs (-/+ DBR1) to YCAA RNPs\n",
    "\n",
    "def matrix_from_bam(bampath, logofrom = 263, logolength=30):\n",
    "    samfile = pysam.AlignmentFile(bampath, \"rb\")\n",
    "    n=0\n",
    "    matrix = {'A':np.zeros(logolength),'C':np.zeros(logolength), 'G':np.zeros(logolength), 'T':np.zeros(logolength)}\n",
    "    primers = []\n",
    "\n",
    "    for read in samfile.fetch(contig='pLG128',start=480,stop=500): # fetch reads mapping to end of ncRNA\n",
    "        if read.query_length < 270: # just take reads spanning full ncRNA\n",
    "            continue\n",
    "        if not read.is_forward: #discard any reverse reads (only 0.09%)\n",
    "            continue\n",
    "        UMI = read.seq[0:6]\n",
    "\n",
    "        n+=1\n",
    "        try:\n",
    "            print_from = read.get_reference_positions().index(logofrom + 227)\n",
    "        except:\n",
    "            continue\n",
    "        if read.seq[-1] == 'A': # for this particular experiment, the ddA tail added by BoMoC during sequencing was not trimmed. Do this now.\n",
    "            sequence = read.seq[print_from:-1]\n",
    "        else:\n",
    "            sequence = read.seq[print_from:]\n",
    "        for i,x in enumerate(sequence):\n",
    "            if i >= logolength:\n",
    "                continue\n",
    "            matrix[x][i]+=1\n",
    "        \n",
    "        primers.append(sequence)\n",
    "    return matrix, primers\n",
    "\n",
    "\n",
    "def freq_matrix_from_seq_list(seq_list, logolength=10, normalize = False):\n",
    "    matrix = {'A':np.zeros(logolength),'C':np.zeros(logolength), 'G':np.zeros(logolength), 'T':np.zeros(logolength)}\n",
    "    totalcounts = 0\n",
    "    for seq in seq_list:\n",
    "        totalcounts +=1\n",
    "        for i,b in enumerate(seq):\n",
    "            if i >= logolength:\n",
    "                continue\n",
    "            if b == '_': # used to notate balnk:\n",
    "                continue\n",
    "            matrix[b][i]+=1\n",
    "    if normalize:\n",
    "        matrix_norm = {k:v/totalcounts*100 for (k,v) in matrix.items()}\n",
    "        return matrix_norm\n",
    "    else:\n",
    "        return matrix\n",
    "\n",
    "bampaths = [\"OTTR_reads/sam/A5_to_LG128_geneious.bam\",\n",
    "            \"OTTR_reads/sam/A6_to_LG128_geneious.bam\",\n",
    "            \"OTTR_reads/sam/A7_to_LG128_geneious.bam\"]\n",
    "\n",
    "matrices = [matrix_from_bam(path, logofrom=268, logolength=19) for path in bampaths]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this cell parses the primer-binding site mutant library. It creates the list \"pairs\" where each member of the list is a tuple of the PBS sequence and the linked primer sequence.\n",
    "# it also applies some filters so that only the really good reads are used.\n",
    "\n",
    "good_reads = 0\n",
    "all_reads = 0\n",
    "pysam.sort(\"-o\", \"tmp.bam\", 'OTTR_reads/sam/H11.sam')\n",
    "pysam.index(\"tmp.bam\")\n",
    "samfile = pysam.AlignmentFile('tmp.bam', 'rb')\n",
    "\n",
    "pairs = []\n",
    "\n",
    "for read in samfile.fetch(contig=samfile.references[0],start=480,stop=500): # fetch reads mapping to end of ncRNA\n",
    "    all_reads += 1\n",
    "    if read.is_forward: #discard any \"reverse\" reads. This sequencing run used read 2, not read 1, so \"read.is_forward\" means we don't want it\n",
    "        continue\n",
    "    if read.query_length != 270: # this is the maximum readlength after adapter trimming, and means most reads should correspond to the full ncRNA\n",
    "        continue\n",
    "\n",
    "    reference_positions = read.get_reference_positions()\n",
    "    if not ((285 in reference_positions) and (400 in reference_positions) and (450 in reference_positions)): # just checking that several positions throughout the ncRNA are mapped\n",
    "        continue\n",
    "\n",
    "    try:\n",
    "        PBS_start = reference_positions.index(367)\n",
    "        PBS_end = reference_positions.index(375)\n",
    "        if PBS_end-PBS_start != 8:\n",
    "            continue #avoid indels in PBS\n",
    "\n",
    "        test_sequence_position = reference_positions.index(320)\n",
    "        if read.query_sequence[test_sequence_position:test_sequence_position+20] != 'TTGGATGATATTCATAATTG': # check that the read is sensible and aligned correctly\n",
    "            continue\n",
    "\n",
    "        PBS_seq = read.query_sequence[PBS_start:PBS_end+1]\n",
    "\n",
    "        if read.query_sequence[PBS_start-10:PBS_start] != 'ACATTATACA' or read.query_sequence[PBS_end+1:PBS_end+11] != 'ACATTTGCGG': # check that the sequence around the PBS is perfect\n",
    "            continue\n",
    "\n",
    "        PBS_revcomp = str(Seq(PBS_seq).reverse_complement())\n",
    "        primer_start = reference_positions.index(505) #actually the base before the primer\n",
    "        if read.query_sequence[primer_start-10:primer_start] != 'TCCGGTTTTG': #the sequence before the primer should be correct\n",
    "            continue\n",
    "        primer_seq = read.query_sequence[primer_start:]\n",
    "        UMI=read.seq[0:6]\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "    good_reads+=1\n",
    "    primer_length = len(primer_seq)\n",
    "    pairs.append((PBS_seq, primer_seq))\n",
    "\n",
    "print(good_reads, all_reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# YCAA experiment, and select examples from the PBS library, for main figure\n",
    "\n",
    "fig, axs = plt.subplots(nrows=8, figsize=(5,5))\n",
    "\n",
    "\n",
    "WTpbs = 'GCATATCTA'\n",
    "\n",
    "for i in range(8):\n",
    "        if i > 2:\n",
    "            query_pbs = ['GCATATATA','GCATAACTA', 'GCATGTCTA','GCAGATCTA','GCGTATCTA'][i-3] # the list of PBS mutants chosen for the main figure\n",
    "            primers = [p[1] for p in pairs if p[0]==query_pbs] \n",
    "        else:\n",
    "            primers = [p[10:] for p in matrices[i][1] if len(p)>9]\n",
    "        \n",
    "        matrix = freq_matrix_from_seq_list(primers, logolength=10)\n",
    "        matrix = {k:v/1000 for (k,v) in matrix.items()}\n",
    "\n",
    "        ax=axs[i]\n",
    "        \n",
    "        if i < 3:\n",
    "            logo = logomaker.Logo(pd.DataFrame(matrix), ax=ax, color_scheme=logo_colorscheme, font_name='Helvetica')\n",
    "        else:\n",
    "            logo = logomaker.Logo(pd.DataFrame(matrix), ax=ax, color_scheme=greypalette[2], font_name='Helvetica')\n",
    "            logo.style_glyphs_in_sequence(sequence='CTAGATATGC', color=greypalette[0])\n",
    "            logo.style_single_glyph(p=[3,4,5,6,7][i-3],c='TTCCC'[i-3],color=logo_colorscheme['TTCCC'[i-3]])\n",
    "        ax.set_xticks(range(1+len(WTpbs)))\n",
    "        ax.get_xaxis().set_visible(False)\n",
    "axs[-1].get_xaxis().set_visible(True)   \n",
    "axs[-1].set_xticklabels('A'+WTpbs[::-1])\n",
    "\n",
    "plt.savefig(\"Fig4D.pdf\", transparent=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fig S6B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(nrows=4, ncols=9,figsize=(24,5))\n",
    "\n",
    "WTpbs = 'GCATATCTA'\n",
    "for i in range(9):\n",
    "    for j in range(4):\n",
    "        pbs = WTpbs[0:i] + 'ACGT'[j] + WTpbs[i+1:]\n",
    "        if 'ACGT'[j] == WTpbs[i]:\n",
    "            WT=True\n",
    "        else:\n",
    "            WT=False\n",
    "        primers = [p[1] for p in pairs if p[0]==pbs]\n",
    "        matrix = freq_matrix_from_seq_list(primers, normalize=False)\n",
    "        ax=axs[j,8-i]\n",
    "        logo = logomaker.Logo(pd.DataFrame(matrix), ax=ax, color_scheme=greypalette[2], font_name='Helvetica')\n",
    "        logo.style_glyphs_in_sequence(sequence='CTAGATATGC', color=greypalette[0])\n",
    "        logo.style_single_glyph(p=9-i,c='TGCA'[j],color=logo_colorscheme['TGCA'[j]])\n",
    "        ax.set_xticks(range(1+len(pbs)))\n",
    "        ax.set_xticklabels('A'+pbs[::-1])\n",
    "        if WT:\n",
    "            ax.get_xticklabels()[9-i].set_fontweight('heavy')\n",
    "        else:\n",
    "            ax.get_xticklabels()[9-i].set_color('r')\n",
    "\n",
    "fig.tight_layout(pad=0.5)\n",
    "plt.savefig(\"FigS6B.pdf\", transparent=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fig S6C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samfile = pysam.AlignmentFile('OTTR_reads/sam/H12_trim4_to_pLG128.bam', \"rb\")\n",
    "samfile_OH = pysam.AlignmentFile('OTTR_reads/sam/H11_trim4_to_pLG128.bam', \"rb\")\n",
    "\n",
    "n=0\n",
    "max_length = 30\n",
    "primers_dNTPs = []\n",
    "pad  = 7\n",
    "\n",
    "sams = [(samfile,'R2'),(samfile_OH,'R1')] # the no-OH sample dervies from Read 1 data, the +OH sample derives from Read 2 data\n",
    "outs = [[],[]]\n",
    "\n",
    "# For this experiment, because dNTPs were added, there's quite a bit ccDNA around, which can either map to the end of the ncRNA (which it is linked to), or to the reverse complement of the template region.\n",
    "# We want to visualize the reads corresponding to the 3' end of the RNA. These either map in forward orientation to 3' end of RNA (if read 1), or in reverse orientation to the PBS (if read 1).\n",
    "# (and map the other way around if read 2 data)\n",
    "\n",
    "for i, (samfile, readdir) in enumerate(sams):\n",
    "    for read in samfile.fetch(contig='pLG128'):\n",
    "        if (read.is_reverse and readdir=='R2') or (read.is_forward and readdir=='R1'):\n",
    "            # get plus-strand reads mapping to the end of the ncRNA\n",
    "            try:\n",
    "                primer_start = read.get_reference_positions().index(508)\n",
    "            except:\n",
    "                continue\n",
    "            sequence = read.seq\n",
    "        else:\n",
    "            # get minus-strand reads mapping to the PBS\n",
    "            try:\n",
    "                primer_start = read.get_reference_positions().index(373)\n",
    "            except:\n",
    "                continue\n",
    "            sequence = str(Seq(read.seq).reverse_complement())\n",
    "            primer_start = len(sequence) - primer_start - 1\n",
    "\n",
    "        print_from = primer_start - pad\n",
    "        if print_from < 0:\n",
    "            padding = '_'*(print_from * -1)\n",
    "            outs[i].append(padding+sequence[0:max_length+print_from])\n",
    "        else:\n",
    "            outs[i].append(sequence[print_from:print_from+max_length])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(figsize=(8,len(outs)),nrows=len(outs),sharex=True)\n",
    "for i,out in enumerate(outs):\n",
    "    plotme=freq_matrix_from_seq_list(out, logolength=20)\n",
    "    logomaker.Logo(pd.DataFrame(plotme), ax=axs[i], color_scheme=logo_colorscheme, font_name='Helvetica')\n",
    "    axs[i].set_xticks(range(20))\n",
    "axs[-1].set_xticklabels('GGACAAUCUAUACGACAUAU')\n",
    "plt.savefig(\"FigS6C.pdf\", transparent=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 4G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the distribution of P-O3' distances in RNA. Using the DRT2 RNP resting state itself to derive these distributions. It has a good mix of A-form helices and other linkers, so should give a reasonably representative distribution.\n",
    "\n",
    "from Bio.PDB import *\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure('DRT2', 'DRT2rnp.pdb')\n",
    "Ps = []\n",
    "Os = []\n",
    "indices = []\n",
    "\n",
    "for model in structure:\n",
    "    for residue in model['R']:\n",
    "        try:\n",
    "            Ps.append(np.array(residue['P'].get_coord()))\n",
    "            Os.append(np.array(residue[\"O3'\"].get_coord()))\n",
    "            indices.append(residue.get_id()[1])\n",
    "        except:\n",
    "            continue\n",
    "Ps = np.array(Ps)\n",
    "Os = np.array(Os)\n",
    "indices = np.array(indices)\n",
    "\n",
    "fullPs = np.zeros((280,3))\n",
    "for n,index in enumerate(indices):\n",
    "    fullPs[index,:] = Ps[n]\n",
    "fullPs = np.where(fullPs==0, np.nan, fullPs) # inserting nans where the chain breaks, so we don't measure distances across chain breaks.\n",
    "\n",
    "fullOs = np.zeros((280,3))\n",
    "for n,index in enumerate(indices):\n",
    "    fullOs[index,:] = Os[n]\n",
    "fullOs = np.where(fullOs==0, np.nan, fullOs) # ditto\n",
    "\n",
    "distances = []\n",
    "for i in range(277):\n",
    "    distances.append(np.sqrt(((fullPs[i+3] - fullOs[i])**2).sum())) # get all distances between 3'OHs and phosphate 2 bases away\n",
    "\n",
    "fig,ax = plt.subplots(figsize=(3,1))\n",
    "ax.hist(distances,bins=20,color=drtpalette[3])\n",
    "ax.axvline(11.7, ls='--', c='k') # to indicate the observed distance beween the primer 5'P and C278 3'O\n",
    "\n",
    "ax.set_ylabel('frequency')\n",
    "ax.set_xlabel('distance (A)')\n",
    "plt.savefig(\"Fig4G.pdf\", transparent=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioconda2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
